solidity/contracts/bridge/Bridge.sol

```diff
@@ -141,36 +180,165 @@ contract Bridge {
-        //      amount to update their Bank balance, minus fee.
-        //
-        // TODO We need to validate if the sum in the output minus the
-        //      amount from the previous wallet balance input minus fees is
-        //      equal to the amount by which Bank balances were increased.
-        //
-        // TODO We need to validate txOutput to see if the balance was not
-        //      transferred away from the wallet before increasing balances in
-        //      the bank.
-        //
-        // TODO Delete deposit from unswept mapping or mark it as swept
-        //      depending on the gas costs. Alternativly, do not allow to
-        //      use the same TX input vector twice. Sweep should be provable
-        //      only one time.
+        validateSweepProof(
+            txVersion,
+            txInputVector,
+            txOutputVector,
+            txLocktime,
+            merkleProof,
+            txIndexInBlock,
+            bitcoinHeaders
+        );
+        updateBalances(txInputVector, txOutputVector);
     }
 
     // TODO It is possible a malicious wallet can sweep deposits that can not
     //      be later proved on Ethereum. For example, a deposit with
     //      an incorrect amount revealed. We need to provide a function for honest
     //      depositors, next to sweep, to prove their swept balances on Ethereum
     //      selectively, based on deposits they have earlier received.
+
+    function updateBalances(
+        bytes memory txInputVector,
+        bytes memory txOutputVector
+    ) internal {
+        // We need to read `fundingTxHash`, `fundingOutputIndex` and
+        // P2SH script depositor address from `txInputVector`.
+        // We then hash them to obtain deposit identifier and read
+        // DepositInfo. From DepositInfo we know what amount was declared
+        // by the depositor in their reveal transaction and we use that
+        // amount to update their Bank balance, minus fee.
+        uint256 varIntDataLength;
+        uint256 inputCounter;
+        (varIntDataLength, inputCounter) = BTCUtils.parseVarInt(txInputVector);
+        uint256 offset = 1 + varIntDataLength;
+
+        for (uint256 i = 0; i < inputCounter; i++) {
+            bytes memory input =
+                txInputVector.slice(offset, txInputVector.length - offset);
+
+            bytes32 fundingTxHash = input.extractInputTxIdLE();
+            bytes memory fundingOutputIndex = input.extractTxIndexLE();
+            bytes memory scriptSignature = input.extractScriptSig();
+
+            // Script signature has the following format:
+            // <signature> <public key> <redeemScript>.
+            // The depositor address is 20 bytes long and is stored in the
+            // `redeemScript`. The data that follows the depositor address has
+            // constant length of 91 bytes.
+            address depositorAddress =
+                scriptSignature.toAddress(scriptSignature.length - 91);
```

Comment:
`91`will be no longer true if changes from #90 land on `main`. Please post a TODO to not miss that.

Reply:
`91` already takes into account that we added hashes instead of values in the redeem script. It seems to be working.

Reply:
Ah ok, I misunderstood how `toAddress` works. `91` is correct, you're right.